16.0 mig custom report

Honeyxilia opened the pull request.

fcayre committed:
Import the custom_report module into the 12.0 branch

legalsylvain committed:
[ADD] migration analysis files

fcayre committed:
Fix tests and dependencies

fcayre committed:
More trivial fixes: model and attribute names, string dates and the like

fcayre committed:
Catchup with commown_lead_risk_analysis last 10.0 devs

fcayre committed:
Remove no more existing recurring_invoices field from custom_report tests

fcayre committed:
Fixup ddc60124a7b9603581ad0a6418cd0ca9c2fbb84a

Some code was removed by error in custom_report.

fcayre committed:
Reactivate CI on custom_report

...  thanks to new account_usability fork, with setuptools config included.

fcayre committed:
Fix copyright and such

legalsylvain committed:
[REF] apply pre-commit, fixing website values

fcayre committed:
Fix report and tests wrt 12.0 API changes

In the report:
- analytic_account_id has been replaced by contract_line_id.contract_id on invoice lines
- dict.values() must be replaced by list(dict.values()) (py3 related change)

In the tests:
- selected partner's xml id was no more a b2c partner
- contract and invoice are now related by their lines.

legalsylvain committed:
[IMP] pre-commit : apply trailing-whitespace

fcayre committed:
Fix custom_report tests to make them work with website_sale_affiliate installed

fcayre committed:
Move mocked session test mixin class setup into instance setup

Otherwise concrete test classes that implement it are not
initialized with their env when we try to modify it and
the class init crashes.

fcayre committed:
Invoice list on portal is now a direct download page

We do not want to maintain an html invoice template and a py3o template,
so the invoice page is not easily accessed anymore (unless you know the
direct URL, and this does not hurt anyway).

fcayre committed:
Add custom py3o report data for sale order and fix invoice ones

fcayre committed:
Adapt sale order report to v12 and add a simple unittest

The test does not check anything else than there is no error yet.

legalsylvain committed:
[META] enable black

legalsylvain committed:
[META] enable isort

legalsylvain committed:
[META] enable black (new version)

fcayre committed:
Remove bottom notes on PDF invoices

This is done by removing the corresponding table in the template (based
on the "comment" field on the invoice object).

fcayre committed:
Fix description in manifest (invalid ReST)

github-grap-bot committed:
custom_report 12.0.1.0.1

fcayre committed:
Use analytic account instead of (lost) contract line to display contract on name on lines of the invoice PDF report

github-grap-bot committed:
custom_report 12.0.1.0.2

fcayre committed:
Restore report t10n feature for invoices

fcayre committed:
Add t10n support for sale order py3o report

fcayre committed:
Refactor tests in a view to add sale order report ones

fcayre committed:
More t10n

fcayre committed:
Add tests for the sale order report

fcayre committed:
Small de t10n adjustments

fcayre committed:
Translate 'Page' in invoice and order PDF reports

github-grap-bot committed:
custom_report 12.0.1.0.3

fcayre committed:
Enlarge the qty column of the invoice PDF custom py3o report

github-grap-bot committed:
custom_report 12.0.1.0.4

fcayre committed:
Add missing i18n pot files

fcayre committed:
Automatically rename fields with script

Script is:
```

rename() {
    oldword=$1
    newword=$2

    git grep -l $oldword | grep -vE '\.po(t)?$' | xargs sed -i "s@$oldword@$newword@g"
}

rename is_rental has_recurrent_payment
rename rental_price_extra recurrent_payment_amount_extra
rename rental_price recurrent_payment_amount
rename rental_tax_ids recurrent_payment_tax_ids
rename rental_frequency recurrent_payment_frequency
rename rental-ratio recurrent-payment-ratio
```

fcayre committed:
i18n update of all modules impacted by the field renamings

fcayre committed:
Post-merge updates

fcayre committed:
Fix invoice report is_b2c test

The is_b2c test was checking whether the parent was a company instead of the commercial_partner_id,
which yields a different result in case of a delivery address or the like (when the address is not
a direct child of the company).

github-grap-bot committed:
[BOT] post-merge updates

Honeyxilia committed:
[IMP] custom_report: black, isort, prettier

Honeyxilia committed:
[MIG] custom_report: Migration to 16.0

Honeyxilia committed:
[test] Replaced at_install/post_install decorators by a tagged decorator

Honeyxilia committed:
[test] Removed removed mock class from product_rental

Honeyxilia committed:
[test] Replaced setUp by setUpClass

Honeyxilia committed:
[test] Adapted helper function to render method (from ir.action.report) changes

- The method has been renamed from render to _render
- The method requires an additionnal argument : the reference of the report

Honeyxilia committed:
Adapted _render_qweb_pdf method to changes from ir.actions.report models

- The render methods have had an underscore (_) added in front of them
- The render methods now require the reference of the report - now, the actual report is searched in the method itself, in case the self ir.actions.report is another report or is empty.

Honeyxilia committed:
Changed account.invoice to account.move

Honeyxilia committed:
Changed migrated fields in account_invoice.odt

- partner_id.image -> partner_id.image_512
- type -> move_type
- date_invoice -> date
- account_analytic_id -> analytic_line_ids.account_id
- uom_id -> product_uom_id
- ('open', 'paid') -> 'posted'
- Retriving taxes group data rather than individual taxes (see notes)
- move_name/number -> name
- amount_by_group -> tax_totals['groups_by_subtotal']
- refund_invoice_id -> reversal_move_id
- payment_term_id -> invoice_payment_term_id
- Added html_sanitize around invoice_payment_term_id.note

-- Taxes notes --

- In Odoo 12.0, there was a field in account.invoice (tax_ids) which contained all taxes applying to the items, and contained both the base amount on which the tax is applied, and the resulting tax amount.
- The tax_ids field has been moved to the lines of an account.move, and account.move now has a computized field which gives the base amount and tax amount for each tax group (ie. 0%, 20%,...) but not for each individual tax.
- If we were to retrieve the amount of tax for each individual tax, we can use the inserted tax lines which correspond to each individual tax - however, there are no tax lines if the amount of taxes is 0.

=> After internal discussions, we decided that to display all of the tax groups even if each individual type of tax (goods, services) is not displayed.

Honeyxilia committed:
Changed report text translation fetch method

Since the ir.translation model is no longer available, we change the way we fetch the translation of any given text by using the `get_python_translations` method.

Honeyxilia committed:
[test] Fixed deposit account retrival

- Depending on the test environnement, the env. company might not have an account_chart template set, meaning the accounts listed in l10n_fr might not be initialized for that company - so, we make sure the company's accounts are initialized correctly
- Rather than using a reference to the deposit account, which contains the company id, we use a more robust retrival by searching an account with the deposit code (275100), as well as the current env. company's id

Honeyxilia committed:
[test] Added now required plan_id value to analytic account creation

There are 3 available analytic plans available in the demo data, and I picked the Projects plan arbitrarily

Honeyxilia committed:
[test] Replaced analytic accound fields by analytic_distribution

Honeyxilia committed:
[test] Removed confirmation_date assignment in test

confirmation_date has been removed from the sale.order model, fused into the date_order field, which now has a different meaning based on the state :
"Creation date of draft/sent orders,
Confirmation date of confirmed orders." (meaning state == "sale", confirmed at the start of the test)

So, we simply use the date_order field to check the confirmation date.

Honeyxilia committed:
Changed migrated fields in sale_order.odt

- amount_by_group -> tax_totals['group_by_subtotal']
- image -> image_512
- Removed confirmation_date (fused into date_order)
- Added HTML sanitization around sale_order.note and sale_order.payment_term_id.note (since they were turned from Text to Html)

Honeyxilia committed:
[test] Added test for Notes section in a sale.order report

Honeyxilia committed:
[test] Fixed CI test packages installation for odoo_usability modules

See this commit for more explaination : https://github.com/commown/commown-odoo-addons/commit/ef508717420a9fd2f8244fd033b8b630b51cb6e6

Honeyxilia committed:
[test] Added check for payment terms in account.move report

The presence of the payment terms is test in a sale.order report test, but not in the account.move tests, which can potentially lead to it not being present in the rendered invoice - so, we added this test to check the presence of the Payment terms.

Honeyxilia committed:
[test] Improve cov.: Added no cover flag on debug check

This value is mainly used to output the HTML contents of a report in an outside file for debug reasons - there's no functionnal impact on production and so we add this flag to remove these lines from coverage

Honeyxilia committed:
[test] Added tests for view print actions

Honeyxilia committed:
[test] Improve cov.: Added no cover flags for text translation declaration

The file i18n.py doesn't contain any functionnal code, and is instead used to set translatable fields for the reports - so, its test coverage is useless

Honeyxilia committed:
Improve cov.: Inlined a conditionnal attribution

Honeyxilia committed:
Added taxes total value to account_invoice.odt

Beforehand, we used to display the amount of taxes per group in the summary with the total amount, but not the total amount of taxes.
So, we changed the document and added a dedicated translatable field.

9184Lucky commented on custom_report/tests/test_account_invoice_report.py:
```diff
@@ -20,7 +20,21 @@ def setUpClass(cls):
+
+        deposit_account = cls.env["account.account"].search(
+            [
+                ("code", "=", 275100),
```
Ici dans le message j'aurais changé searching an account with the deposit **id** en searching an account with the deposit **code** pour éviter la confusion avec l'id de l'objet !

Honeyxilia replied:
appliqué c;

9184Lucky commented on custom_report/tests/test_account_invoice_report.py:
```diff
@@ -99,18 +99,18 @@ def sale(self, partner, products):
 
     def open_invoice(self, so, is_refund=False, contract=None):
-        inv = self.env["account.invoice"].browse(so.action_invoice_create())
+        inv = so._create_invoices()
```
Pas hyper important mais pour moi tout le reste du commit sauf ce changement va plutôt dans le commit " Changed account.invoice to account.move"

Honeyxilia replied:
fait c:

9184Lucky commented on custom_report/tests/test_account_invoice_report.py:
```diff
@@ -234,3 +236,12 @@ def test_b2c_qty_zero(self):
         self.assertEqual(
             _product_descriptions(doc2), [self.std_product.name]
```
Je comprends pas le début du message de commit: "The presence of the payment terms is present" :/
Et à la fin peut être changer "Payment terms are well present" par "Payment terms are **indeed** present" ça me semble plus idiomatique (mais c'est pas très important...

Honeyxilia replied:
Pour le premier point, c'était une erreur de distraction : je souhaitais dire "the presence of the payment terms is tested in a sale.order report test"

Pour le second point, j'ai totalement reformulé c:

Honeyxilia committed:
[test] Changed account.move creation method from sale.order

Honeyxilia committed:
[test] Added taxes information check for account.move reports

We verify the taxes in two situations :
- An invoice with a 0% tax - in this situation, the tax table should show, but no taxes lines should be displayed in the invoice recap (which is the case currently in 12.0)
- An invoice with a 10% tax - in this situation, both the tax table and the tax recap should be displayed.

fcayre commented on custom_report/tests/test_account_invoice_report.py:
```diff
@@ -20,7 +20,21 @@ def setUpClass(cls):
         cls.b2b_partner = cls.partner = cls.env.ref("base.res_partner_address_1")
 
-        deposit_account = cls.env.ref("l10n_fr.1_pcg_2751")
+        if not cls.env.company.chart_template_id:  # pragma: no cover
```
Ce module dépend de l10n_fr, je pense donc que cette section est inutile, à vérifier.

Honeyxilia replied:
C'est necessaire en local car la compagnie d'environnement n'a pas ces comptes de base
on peut voir pour utiliser uniquement le compte créé dans l10n_fr cependant en effet

Honeyxilia replied:
J'ai essayé de retirer la regénération des comptes, cependant au moment de la création de l'entrée `account.move` (`_create_invoice`), il échoue car il n'y a pas de journal associé au compte de l'environnement actuel (en local)

fcayre replied:
Pas grave, on va pas perdre plus de temps là-dessus même s'il y a sûrement une solution plus élégante, OSEF pour le moment.

fcayre commented on custom_report/models/ir_actions_report.py:
```diff
@@ -6,13 +6,14 @@ class IrActionsReport(models.Model):
     _inherit = "ir.actions.report"
 
-    def render_qweb_pdf(self, res_ids=None, data=None):
+    def _render_qweb_pdf(self, report_ref, res_ids=None, data=None):
```
Qu'est-ce qui empêche de vérifier que cette méthode ne plante pas et retourne un fichier PDF valide ?
(si besoin en utilisant PyPDF2 comme dans commown_shipping)

Honeyxilia replied:
Fait c:

Honeyxilia committed:
[test] Added test for _render_qweb_pdf rendering py3o reports

The _render_qweb_pdf is a method used in some views to render reports that are initially declared as qweb-pdf, but that we overload as py3o reports. So, we overloaded the _render_qweb_pdf method in order to correctly render the now-py3o reports through the _render_py3o method (see custom_report/models/ir_actions_report.py).
To ensure that the overloaded method doesn't encounter any issues for both qweb-pdf and py3o reports, and that the returned files are pdf's, we wrote this test testing both the base behavior (qweb-pdf) and the introduced alternative behavior (py3o).

fcayre commented:
/ocabot rebase

fcayre commented:
/ocabot merge patch
